include/libopencm3/nrf/common/periph.h

```diff
@@ -0,0 +1,148 @@
+ * task having a separate register in that peripheral's task register group.
+ *
+ * A task is triggered when firmware writes a '1' to the task register or when
+ * the peripheral itself, or another peripheral, toggles the corresponding task
+ * signal.
+ */
+
+/** Starting address of all the tasks in the peripheral. */
+#define PERIPH_TASK_OFFSET         (0x000)
+
+/*
+ * Events are used to notify peripherals and the CPU about events that have
+ * happened, for example, a state change in a peripheral. A peripheral may
+ * generate multiple events with each event having a separate register in that
+ * peripheral’s event register group.  An event is generated when the
+ * peripheral itself toggles the corresponding event signal, whereupon the
+ * event register is updated to reflect that the event has been generated.
+ */
+
+/** Starting address of all the events in the peripheral. */
+#define PERIPH_EVENT_OFFSET        (0x100)
+
+#define PERIPH_TRIGGER_TASK(task)  (task) = (1)
+
+/* All peripherals on the APB bus support interrupts. A peripheral only
+ * occupies one interrupt, and the interrupt number follows the peripheral ID,
+ * for example, the peripheral with ID=4 is connected to interrupt number 4 in
+ * the Nested Vector Interrupt Controller (NVIC).
+ */
+
+#define PERIPH_ENABLE_IRQ(base)    nvic_enable_irq(periph_id_from_base(base))
+#define PERIPH_DISABLE_IRQ(base)   nvic_disable_irq(periph_id_from_base(base))
+
+/* Common regisgers. Not all peripherals have these registers, but when they
+ * are present, they are at this offset.
+ */
+#define PERIPH_SHORTS_OFFSET       (0x200)
+#define PERIPH_INTEN_OFFSET        (0x300)
+#define PERIPH_INTENSET_OFFSET     (0x304)
+#define PERIPH_INTENCLR_OFFSET     (0x308)
+
+#define _PERIPH_SHORTS(base)        MMIO32((base) + PERIPH_SHORTS_OFFSET)
+#define _PERIPH_INTEN(base)         MMIO32((base) + PERIPH_INTEN_OFFSET)
+#define _PERIPH_INTENSET(base)      MMIO32((base) + PERIPH_INTENSET_OFFSET)
+#define _PERIPH_INTENCLR(base)      MMIO32((base) + PERIPH_INTENCLR_OFFSET)
+
+/* TODO: convert these to functions */
+#define periph_enable_shorts(base, shorts)    periph_shorts(base) |= (shorts)
+#define periph_disable_shorts(base, shorts)   periph_shorts(base) &= (~(shorts))
+#define periph_clear_shorts(base)             periph_shorts(base) = (0)
+
+#define periph_enable_interrupts(base, mask)     periph_intenset(base) |= (mask)
+#define periph_disable_interrupts(base, mask)    periph_intenclr(base) = (mask)
+#define periph_clear_interrupts(base)            periph_intenclr(base) = (0xffffffff)
+
+/** Mark the signal as not connected to any pin. */
+#define GPIO_UNCONNECTED		0xFFFFFFFFU
+
+/** This is an approximation of log2. As used here, works correctly
+ * only for single bit set, which should be the case when used to.
+ * convert above GPIOxy macros to pin numbers as needed for PSEL
+ * registers of peripherals.
+ */
+#define __GPIO2PIN(x)			(31 - __builtin_clz((uint32_t) (x)))
```

Comment:
hrm, this might surprise people then.  On the other targets, the gpio_ functions can take a bitfield with all pins getting the same config?

Reply:
Yeah, that's why I prefixed it by __ and put it here into periph.h, which shouldn't be explicitly included by the user

Reply:
It's an option to just put it in the .c files themselves, so they don't even ever show up in user space too, fwiw.